src/acanban/auth.py

```diff
@@ -54,40 +56,61 @@ class User(AuthUser):
-        password, name, email, role = current_app.auth_manager.users[username]
         self.username = username
-        self.password = password
-        self.name = name
-        self.email = email
-        self.role = role
+        self.r = r.table('users').get(self.username)
+
+    @property
+    async def password(self) -> str:
+        async with current_app.db_pool.connection() as connection:
+            return await self.r['password-hash'].run(connection)
+
+    @property
+    async def name(self) -> str:
+        async with current_app.db_pool.connection() as connection:
+            return await self.r['name'].run(connection)
+
+    @property
+    async def email(self) -> str:
+        async with current_app.db_pool.connection() as connection:
+            return await self.r['email'].run(connection)
+
+    @property
+    async def role(self) -> str:
+        async with current_app.db_pool.connection() as connection:
+            return await self.r['role'].run(connection)
 
 
 class Authenticator(AuthManager):
-    users: Dict[Optional[str], Tuple[Optional[str], Optional[str],
-                                     Optional[str], Optional[str]]]
     user_class = User
 
     def init_app(self, app: Quart) -> None:
         """Embed auth_manager attribute into app."""
         super().init_app(app)
-        # TODO: wrap the database instead
-        self.users = {None: (None, None, None, None)}
+        self.r = r.table('users')
 
-    def add_user(self, username: str, password: str,
-                 name: str, email: str, role: str) -> None:
+    async def add_user(self, username: str, password: str,
+                       name: str, email: str, role: str) -> None:
         """Try to add user to database."""
-        if username in self.users: raise ValueError('username taken')
         # This never happens through browser, but via manual requests.
         if role not in ROLES: raise ValueError('unknown role')
-        self.users[username] = crypt(password), name, email, role
-
-    def log_user(self, username: str, password: str) -> User:
+        async with current_app.db_pool.connection() as connection:
+            user = await self.r.get(username).run(connection)
+            user = {'username': username, 'password': crypt(password),
+                    'name': name, 'role': role}
+            if await self.r.insert(user).run(connection)['errors']:
+                raise ValueError('username taken')
+
+    async def log_user(self, username: str, password: str) -> User:
         """Try to log the given user in."""
-        digest, *info = self.users[username]
-        assert digest is not None
+        async with current_app.db_pool.connection() as connection:
+            user_query = self.r.with_fields('username', 'password')
+            user = await user_query.get(username).run(connection)
+        assert user is not None
+        user = User(user['username'])
+        digest = await user.password
```

Comment:
Lol I was to focused on the details that I forgot about this.

```suggestion
        user = User(username)
        digest = await user.password
```

Also `User` is not awaitable and we need to handle cases when properties doesn't exists (`rethinkdb.errors.ReqlNonExistenceError`).